### PR TITLE
Add Lanczos interpolation mode for CPU images

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR adds `mode="lanczos"` to `F.interpolate`. **It is bitwise equal to PIL-SIMD on uint8, and 2X-10X faster than PIL on relevant cases**.

Lanczos mode is a popular interpolation mode, often considered of superior quality to bicubic. Supporting it in `torch` is necessary for those users still relying on PIL for their preproc.

The core change is just a new `HelperInterpLanczos `struct with the lanczos filter. The existing separable `interpolate` code is generic enough that we get float and uint8 paths, as well as SIMD-sepcific implem (AVX2 and NEON) for free.

The majority of the changes are pure plumbing + tests.

### What is supported

- CPU only
- uint8 and float dtypes
- [AVX2](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/UpSampleKernelAVXAntialias.h) and [NEON](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/UpSampleKernelNEONAntialias.h) paths
- `antialias=True` only
- image batches only (no 1D or 3D resampling).
- backward pass

                                                                                                              

### What is *not* (yet?) supported

  - CUDA/MPS
  - `antialias=False`. I can follow-up with this one, ideally staying close to opencv.
  - 1D and 3D resampling

### Tests:

**TL;DR** Output on `uint8` dtype is bitwise exact vs PIL-SIMD, and within `atol=2` on PIL, which is the same as for bicubic mode. Results on floats (vs uint8 path) are also very close, with expected differences due to the truncation of intermediate buffers on the uint8 path (again, this is expected, documented, and very similar to bicubic mode).

<details>

```py
import numpy as np
import pytest
import torch
import torch.nn.functional as F
from PIL import Image


def is_pillow_simd():
    """Detect Pillow-SIMD (has .post suffix in version string)."""
    return ".post" in getattr(Image, "__version__", "")


PIL_RESAMPLE = {
    "bilinear": Image.BILINEAR,
    "bicubic": Image.BICUBIC,
    "lanczos": Image.LANCZOS,
}

# (mode, pillow_simd) -> atol for uint8
UINT8_ATOLS = {
    ("bilinear", True): 0,
    ("bilinear", False): 1,
    ("bicubic", True): 0,
    ("bicubic", False): 2,
    ("lanczos", True): 0,
    ("lanczos", False): 2,
}

# (mode,) -> (atol, min_percent_within_atol) for float32
# e.g. (1, 99) means atol=1 must pass on >= 99% of pixels
FLOAT32_ATOLS = {
    "bilinear": (1, 100),
    "bicubic": (2, 97),
    "lanczos": (3, 95),
}


PIL_MODE = {1: "L", 3: "RGB"}


def pil_resize(input_uint8_tensor, output_size, mode):
    """Resize a (C, H, W) uint8 tensor using PIL, return (C, H_out, W_out) uint8 tensor."""
    c, h, w = input_uint8_tensor.shape
    out_h, out_w = output_size
    # CHW -> HWC for PIL
    arr = input_uint8_tensor.permute(1, 2, 0).numpy()
    if c == 1:
        arr = arr.squeeze(-1)
    pil_img = Image.fromarray(arr, mode=PIL_MODE[c])
    pil_out = pil_img.resize((out_w, out_h), resample=PIL_RESAMPLE[mode])
    result = torch.from_numpy(np.array(pil_out))
    if c == 1:
        result = result.unsqueeze(0)
    else:
        # HWC -> CHW
        result = result.permute(2, 0, 1)
    return result


INPUT_SIZES = [(8, 8), (15, 15), (64, 64), (100, 73), (256, 256), (500, 332)]
SCALE_FACTORS = [0.25, 0.5, 0.75, 1.0, 1.5, 2.0, 3.0, 4.0]
CHANNELS = [1, 3]
MEMORY_FORMATS = [torch.contiguous_format, torch.channels_last]
MODES = ["bilinear", "bicubic", "lanczos"]
DTYPES = [torch.uint8, torch.float32]


@pytest.mark.parametrize("mode", MODES)
@pytest.mark.parametrize("dtype", DTYPES, ids=["uint8", "float32"])
@pytest.mark.parametrize("input_size", INPUT_SIZES, ids=lambda s: f"{s[0]}x{s[1]}")
@pytest.mark.parametrize("scale_factor", SCALE_FACTORS)
@pytest.mark.parametrize("num_channels", CHANNELS)
@pytest.mark.parametrize("memory_format", MEMORY_FORMATS, ids=["contiguous", "channels_last"])
def test_interpolate_correctness(mode, dtype, input_size, scale_factor, num_channels, memory_format):
    in_h, in_w = input_size
    out_h = max(1, int(in_h * scale_factor))
    out_w = max(1, int(in_w * scale_factor))

    torch.manual_seed(42)
    x = torch.randint(0, 256, (1, num_channels, in_h, in_w), dtype=torch.uint8)
    x = x.contiguous(memory_format=memory_format)

    if dtype == torch.uint8:
        ref = pil_resize(x[0], (out_h, out_w), mode).unsqueeze(0)
    else:
        # For float, compare against the uint8 torch path (which is validated
        # against PIL in the uint8 test).
        ref = F.interpolate(
            x, size=(out_h, out_w), mode=mode, align_corners=False, antialias=True
        ).float()
        x = x.float() / 255.0

    torch_out = F.interpolate(
        x, size=(out_h, out_w), mode=mode, align_corners=False, antialias=True
    )

    if dtype == torch.uint8:
        atol = UINT8_ATOLS[(mode, is_pillow_simd())]
        torch.testing.assert_close(torch_out, ref, atol=atol, rtol=0)
    else:
        torch_out = (torch_out.clip(0, 1) * 255.0).round()
        atol, min_pct = FLOAT32_ATOLS[mode]
        diff = (torch_out - ref).abs()
        pct_within = (diff <= atol).float().mean() * 100
        assert pct_within >= min_pct, (
            f"Only {pct_within:.1f}% of pixels within atol={atol} "
            f"(expected >= {min_pct}%), max diff={diff.max()}"
        )

```

</details>

### Benchmarks

**TL;DR**:
- **AVX2**: 5X-10X faster than PIL, 1.5X-4X faster than PIL-SIMD
- **NEON**: ~2X faster than PIL (no PIL-SIMD NEON to compare against).

<details>

```
AVX2 against PIL:
===============================================================================================
Resolution Scale    Torch (ms)           PIL (ms)             Speedup   
-----------------------------------------------------------------------------------------------
64x64      x0.25    0.03 +- 0.00         0.06 +- 0.01         1.67x
64x64      x0.5     0.03 +- 0.00         0.07 +- 0.00         2.34x
64x64      x2.0     0.06 +- 0.00         0.23 +- 0.01         3.83x
64x64      x4.0     0.13 +- 0.00         0.74 +- 0.36         5.64x
320p       x0.25    0.14 +- 0.00         0.70 +- 0.02         5.01x
320p       x0.5     0.17 +- 0.00         0.89 +- 0.19         5.32x
320p       x2.0     0.55 +- 0.02         4.59 +- 0.15         8.30x
320p       x4.0     1.37 +- 0.31         16.31 +- 0.38        11.89x
720p       x0.25    1.05 +- 0.02         7.53 +- 0.11         7.18x
720p       x0.5     1.43 +- 0.03         9.75 +- 0.09         6.81x
720p       x2.0     5.39 +- 0.91         55.31 +- 0.58        10.27x
720p       x4.0     27.57 +- 1.71        203.61 +- 12.73      7.39x
1080p      x0.25    2.20 +- 0.09         17.06 +- 0.14        7.74x
1080p      x0.5     3.16 +- 0.24         21.97 +- 0.38        6.95x
1080p      x2.0     12.10 +- 2.84        109.98 +- 2.16       9.09x
1080p      x4.0     60.19 +- 0.52        429.27 +- 1.68       7.13x
4K         x0.25    8.27 +- 0.03         66.96 +- 0.23        8.10x
4K         x0.5     11.58 +- 0.03        85.91 +- 0.20        7.42x
4K         x2.0     86.51 +- 1.36        492.02 +- 8.04       5.69x
4K         x4.0     275.75 +- 7.99       1745.59 +- 8.35      6.33x
===============================================================================================


AVX2 against PIL-SIMD:
===============================================================================================
Resolution Scale    Torch (ms)           PIL (ms)             Speedup   
-----------------------------------------------------------------------------------------------
64x64      x0.25    0.04 +- 0.01         0.07 +- 0.01         2.08x
64x64      x0.5     0.07 +- 0.00         0.08 +- 0.00         1.13x
64x64      x2.0     0.15 +- 0.02         0.17 +- 0.00         1.13x
64x64      x4.0     0.33 +- 0.01         0.40 +- 0.01         1.22x
320p       x0.25    0.27 +- 0.10         0.20 +- 0.02         0.74x
320p       x0.5     0.18 +- 0.00         0.22 +- 0.01         1.28x
320p       x2.0     0.74 +- 0.14         1.80 +- 0.13         2.44x
320p       x4.0     1.44 +- 0.40         6.97 +- 0.35         4.83x
720p       x0.25    1.06 +- 0.03         1.49 +- 0.04         1.40x
720p       x0.5     1.41 +- 0.03         2.00 +- 0.05         1.42x
720p       x2.0     5.43 +- 0.84         21.44 +- 0.11        3.95x
720p       x4.0     26.77 +- 0.09        79.03 +- 0.40        2.95x
1080p      x0.25    2.18 +- 0.05         3.13 +- 0.05         1.43x
1080p      x0.5     3.00 +- 0.03         4.26 +- 0.06         1.42x
1080p      x2.0     11.77 +- 1.91        46.56 +- 0.38        3.95x
1080p      x4.0     60.21 +- 0.98        179.09 +- 1.49       2.97x
4K         x0.25    8.19 +- 0.04         11.83 +- 0.04        1.44x
4K         x0.5     11.62 +- 0.02        16.56 +- 0.10        1.43x
4K         x2.0     86.22 +- 0.18        184.47 +- 0.49       2.14x
4K         x4.0     272.85 +- 2.78       724.14 +- 2.68       2.65x
===============================================================================================


NEON against PIL:
===============================================================================================
Resolution Scale    Torch (ms)           PIL (ms)             Speedup   
-----------------------------------------------------------------------------------------------
64x64      x0.25    0.04 +- 0.00         0.06 +- 0.01         1.66x
64x64      x0.5     0.05 +- 0.00         0.07 +- 0.00         1.50x
64x64      x2.0     0.14 +- 0.00         0.22 +- 0.01         1.56x
64x64      x4.0     0.37 +- 0.01         0.66 +- 0.01         1.82x
320p       x0.25    0.26 +- 0.01         0.69 +- 0.01         2.65x
320p       x0.5     0.46 +- 0.01         0.82 +- 0.01         1.77x
320p       x2.0     2.06 +- 0.01         3.53 +- 0.06         1.71x
320p       x4.0     5.76 +- 0.02         11.72 +- 0.05        2.03x
720p       x0.25    2.41 +- 0.01         7.61 +- 0.04         3.16x
720p       x0.5     4.82 +- 0.01         9.06 +- 0.02         1.88x
720p       x2.0     23.56 +- 0.02        42.25 +- 0.28        1.79x
720p       x4.0     67.22 +- 1.08        141.94 +- 2.72       2.11x
1080p      x0.25    5.21 +- 0.02         17.02 +- 0.05        3.27x
1080p      x0.5     10.55 +- 0.05        20.52 +- 0.05        1.95x
1080p      x2.0     52.20 +- 0.04        96.77 +- 0.12        1.85x
1080p      x4.0     154.41 +- 3.44       316.13 +- 0.41       2.05x
4K         x0.25    20.69 +- 0.07        68.87 +- 0.15        3.33x
4K         x0.5     42.10 +- 0.09        81.94 +- 0.87        1.95x
4K         x2.0     204.47 +- 0.16       365.16 +- 7.39       1.79x
4K         x4.0     586.99 +- 0.48       1256.39 +- 3.25      2.14x
===============================================================================================


```

```py
"""Benchmark PyTorch Lanczos vs PIL Lanczos on uint8 channels-last inputs.

Usage:
    python bench_lanczos.py
"""

import torch
import torch.nn.functional as F
import numpy as np
from time import perf_counter_ns
from PIL import Image


def bench(f, *args, num_exp=10, warmup=1, **kwargs):
    for _ in range(warmup):
        f(*args, **kwargs)
    times = []
    for _ in range(num_exp):
        start = perf_counter_ns()
        f(*args, **kwargs)
        end = perf_counter_ns()
        times.append(end - start)
    return torch.tensor(times).float()


def report_stats(times, unit="ms"):
    mul = {"ns": 1, "µs": 1e-3, "ms": 1e-6, "s": 1e-9}[unit]
    times = times * mul
    std = times.std().item()
    med = times.median().item()
    print(f"  {med = :.2f}{unit} +- {std:.2f}")
    return med, std


def torch_resize(x, out_h, out_w):
    F.interpolate(x, size=(out_h, out_w), mode="lanczos", align_corners=False, antialias=True)


def pil_resize(pil_img, out_h, out_w):
    out = pil_img.resize((out_w, out_h), resample=Image.LANCZOS)
    torch.from_numpy(np.array(out)).permute(2, 0, 1)


RESOLUTIONS = {
    "64x64": (64, 64),
    "320p": (240, 320),
    "720p": (720, 1280),
    "1080p": (1080, 1920),
    "4K": (2160, 3840),
}

SCALE_FACTORS = [0.25, 0.5, 2.0, 4.0]


if __name__ == "__main__":
    torch.set_num_threads(1)
    results = []

    for res_name, (in_h, in_w) in RESOLUTIONS.items():
        for scale in SCALE_FACTORS:
            out_h = max(1, int(in_h * scale))
            out_w = max(1, int(in_w * scale))

            print(f"\n{res_name} ({in_h}x{in_w}) -> {out_h}x{out_w} (x{scale})")

            torch.manual_seed(42)
            x = torch.randint(0, 256, (1, 3, in_h, in_w), dtype=torch.uint8)
            x = x.contiguous(memory_format=torch.channels_last)

            arr = x[0].permute(1, 2, 0).numpy()
            pil_img = Image.fromarray(arr, mode="RGB")

            print("  torch:")
            t_torch, s_torch = report_stats(bench(torch_resize, x, out_h, out_w))
            print("  PIL:")
            t_pil, s_pil = report_stats(bench(pil_resize, pil_img, out_h, out_w))

            speedup = t_pil / t_torch if t_torch > 0 else float("inf")
            print(f"  speedup: {speedup:.2f}x")

            results.append((res_name, scale, t_torch, s_torch, t_pil, s_pil, speedup))

    # Print summary table
    print("\n" + "=" * 95)
    print(f"{'Resolution':<10} {'Scale':<8} {'Torch (ms)':<20} {'PIL (ms)':<20} {'Speedup':<10}")
    print("-" * 95)
    for res_name, scale, t_torch, s_torch, t_pil, s_pil, speedup in results:
        torch_str = f"{t_torch:.2f} +- {s_torch:.2f}"
        pil_str = f"{t_pil:.2f} +- {s_pil:.2f}"
        print(f"{res_name:<10} x{scale:<7} {torch_str:<20} {pil_str:<20} {speedup:.2f}x")
    print("=" * 95)

```

</details>

Authored with Claude.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

### CODE REVIEW

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 1]
```diff
@@ -1531,7 +1614,7 @@ void upsample_separable_1d(
 
   if (input_scalar_type == at::kByte) {
     // This is a special branch to provide uint8 dtype support for bilinear and bicubic modes only
```
Comment (Skylion007) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp]: This comment isn't right anymore, right?

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 2]
```diff
@@ -1416,6 +1416,89 @@ struct HelperInterpCubic : public HelperInterpBase {
...
+  static constexpr int interp_size = 6;
+
+  // Taken from
+  // https://github.com/python-pillow/Pillow/blob/8004234d879254cc354935ad42fbb51b1700925e/src/libImaging/Resample.c#L64-L86
+  template<typename scalar_t>
+  static inline scalar_t sinc_filter(scalar_t x) {
+    if (x == 0.0) {
+      return 1.0;
+    }
+    x = x * c10::pi<scalar_t>;
```
Comment (Skylion007) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp]:
```suggestion
    x *= c10::pi<scalar_t>;

File: torch/nn/functional.py [hunk 3]
```diff
@@ -4923,6 +4930,19 @@ def interpolate(  # noqa: F811
             scale_factors,
         )
 
+    if input.dim() == 4 and mode == "lanczos":
+        if align_corners is None:
+            raise AssertionError("align_corners is unexpectedly None")
```
Comment (albanD) [inline torch/nn/functional.py:4935]: Should this also raise an error if align_corners=True since we don't support it?

**Review by albanD — COMMENTED** [review]
Looks great!

Only once comment about the align_corner=True behavior. Which seems like it will return something but not expected to work right now?

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 4]
```diff
@@ -1416,6 +1416,89 @@ struct HelperInterpCubic : public HelperInterpBase {
...
+    return std::sin(x) / x;
+  }
+
+  template<typename scalar_t>
+  static inline scalar_t aa_filter(scalar_t x) {
+    // Lanczos-3 filter: sinc(x) * sinc(x/3) for |x| < 3
+    x = std::abs(x);
+    if (x < 3.0) {
+      return sinc_filter(x) * sinc_filter(x / 3.0);
```
Comment (Skylion007) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:1439]: Why can't we also add this to CUDA? That would be more compelling to add then?

File: torch/nn/functional.py
Comment (Skylion007) [inline torch/nn/functional.py:4472]: If we ever properly type this, we can add overloads that return never to statically test against invalid combos

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp (hunk 4 above)
Comment (NicolasHug) [reply aten/src/ATen/native/cpu/UpSampleKernel.cpp:1439]: I was thinking of leaving out the CUDA implementation in this PR because it's already quite large, and I wanted to optimize for review time. I can follow-up immediately with the CUDA support if you think it's worth it.

**Review by albanD — APPROVED** [review]
Looks great!
Thanks for the update

**Review by malfet — APPROVED** [review]
@NicolasHug I think you'll just need to add it to build_variables.bzl
